controllers/onload_controller.go

```diff
@@ -182,6 +203,72 @@ func createModule(
 	return module, nil
 }
 
+func (r *OnloadReconciler) createControlPlaneDaemonSet(
+	ctx context.Context, onload *onloadv1alpha1.Onload,
+) error {
+	dsName := onload.Name + cpdsSuffix
+
+	// Create the control plane daemon set only if it has not been created.
+	ds := &appsv1.DaemonSet{}
+	err := r.Get(
+		ctx,
+		types.NamespacedName{
+			Name:      dsName,
+			Namespace: onload.Namespace,
+		},
+		ds,
+	)
+	if err == nil || !apierrors.IsNotFound(err) {
+		return err
+	}
```

Comment:
Am I missing something or is there a logic bug here? returning `err` when `err == nil` is just returning `nil` which presumably we don't want to do in an error path

Reply:
This is "if DS already exists (err is nil) OR there is a genuine error (not IsNotFound) then return either nil (the former case) or the error (the latter case)". That's what I think we do with the module kind handing too, though I see it's expressed differently.

I see how it's convoluted and will split these checks into two with a bit of logging!

Still, this is a temporary approach that must be improved with something like https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/controller/controllerutil#CreateOrPatch to patch the existing DS if it doesn't match the desired spec.

Reply:
Ok thanks, I hadn't considered that the if statement was a combination of two genuinely different cases.
Some logging / comments would be nice

Reply:
Done, please take a look at the new code. 👍